urla958522fc1_spirosikmd_inmdsy_08_tgz-pJ8-TesterAppJ8/src/nl/rug/peerbox/logic/filesystem/FileServer.java

```diff
@@ -0,0 +1,102 @@
+package nl.rug.peerbox.logic.filesystem;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import nl.rug.peerbox.logic.Context;
+import nl.rug.peerbox.logic.Peerbox;
+
+import org.apache.log4j.Logger;
+
+public final class FileServer implements Runnable {
+
+	
+
+	private static final Logger logger = Logger.getLogger(FileServer.class);
+	private final ExecutorService pool = Executors.newFixedThreadPool(5);
+	private final Context ctx;
+	boolean alive = true;
+	
+	public FileServer(Context ctx) {
+		this.ctx = ctx;
+	}
+	
+	@Override
+	public void run() {
+		logger.info("Starting server " + ctx.getLocalPeer().getPort());
+		try (ServerSocket server = new ServerSocket()) {
+			server.bind(new InetSocketAddress(ctx.getLocalPeer().getAddress(), ctx.getLocalPeer().getPort()));
+			while (!Thread.currentThread().isInterrupted()) {
+				logger.info("Waiting for incoming connection");
+				try {
+					final Socket s = server.accept();
+					logger.debug("Accepted incoming connection from "
+							+ s.getRemoteSocketAddress());
+					pool.execute(new SendFileTask(s));
+				} catch (IOException e) {
+					logger.error(e);
+				}
+			}
+			pool.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)